apps/labs/posts/building-scipy-with-flang.md

```diff
@@ -0,0 +1,641 @@
+freight train, all we could try to do was cross our fingers that _one_ Fortran
+compiler would be ready in time (most likely either llvm-flang or LFortran), and
+be as prepared as possible in terms of having all the other pieces ready to go.
+LLVM is a big baby, and keeping all the pieces involved up-to-date is a
+non-trivial exercise already[^3], but we put extra effort into trying to
+identify potential problems with Flang early on, both in our own
+[infrastructure](https://github.com/conda-forge/flang-feedstock/pull/27),
+[as well as](https://github.com/llvm/llvm-project/issues/60730)
+[upstream](https://github.com/llvm/llvm-project/issues/63712).
+
+But no matter how motivated we were to pursue this, the iron-clad constraints
+of FOSS remain at play: no-one can take care of everything, much less keep all
+the required knowledge in their head.
+That means we need other people's input & feedback – and when everyone's a
+volunteer[^4], things can take a while.
+
+As it happened, the release of LLVM 17.0 was coming up, which would be the first
+release where Flang had matured enough to remove the `-flang-experimental-exec`
+flag, and was [estimated](https://discourse.llvm.org/t/proposal-rename-flang-new-to-flang/69462/33)
+to be roughly "0.8 level maturity".
+As soon as we had the release candidates built in conda-forge, we tried to throw
+it at SciPy, only to realise that Meson did not support the new Flang yet.
+
+Even worse, it wasn't clear which ABI Flang was using (as mentioned above,
+conda-forge uses the default ABI on Windows everywhere), not least because the
+Clang driver which Flang uses under the hood can target both modes.
+Unfortunately, it wasn't clear which mode was being chosen, and generally it
+looked like Windows support was (as often happens in FOSS) the least mature
+platform.
+
+To top it off, there's also two different standard libraries, two different
+runtime libraries (one of which comes in 4 flavours) and two different linkers
+to consider in all this.
+At this point, we still hadn't compiled more than a "Hello World!" example with
+Flang, so it wasn't even clear that it could compile all of SciPy, much less
+pass the test suite...
+
+## Eucatastrophe
+
+The final push came once SciPy got [ready](https://github.com/scipy/scipy/pull/19317)
+to remove support for `setup.py`, the last link to the "old world" before the
+removal of `distutils`.
+With some kind hints from the Meson devs and a bit of persistence, we managed to
+teach Meson to be able to handle llvm-flang just enough to compile something.
+We eventually passed the first build, and after some more wrangling with Meson,
+the first installation too.
+
+What happened next felt like nothing short of magical – while we were expecting
+test failures, hangs or even crashes, the test suite… just passed 100%?!
+
+!['Screenshot of the first passing test suite with the new flang-built SciPy'](/posts/building-scipy-with-flang/scipy_flang_first_green_CI.png)
```

Comment:
Would this look better as a code block? If you want to keep it like this, also okay with me.

Reply:
It didn't look good in the draft on google docs, but it does look better when rendered 👍